langwatch/src/server/api/routers/triggers.ts

```diff
@@ -0,0 +1,112 @@
+import { z } from "zod";
+import { createTRPCRouter, protectedProcedure } from "../trpc";
+import { TriggerAction } from "@prisma/client";
+import { TRPCError } from "@trpc/server";
+
+import { nanoid } from "nanoid";
+import { TeamRoleGroup, checkUserPermissionForProject } from "../permission";
+
+export const triggerRouter = createTRPCRouter({
+  create: protectedProcedure
+    .input(
+      z.object({
+        projectId: z.string(),
+        name: z.string(),
+        members: z.string().array(),
+        action: z.nativeEnum(TriggerAction),
+        actionParams: z.any(),
+        filters: z.any(),
+        organizationId: z.string(),
+      })
+    )
+    .use(checkUserPermissionForProject(TeamRoleGroup.ALERTS_MANAGE))
+    .mutation(async ({ ctx, input }) => {
+      const organizationUsers = await ctx.prisma.organizationUser.findMany({
+        where: {
+          organizationId: input.organizationId,
+        },
+        include: {
+          user: true,
+        },
+      });
+
+      const organizationUserEmails = organizationUsers.map(
+        (user) => user.user.email
+      );
+
+      input.members.map((email) => {
+        if (!organizationUserEmails.includes(email)) {
+          throw new TRPCError({
+            code: "BAD_REQUEST",
+            message: "Error with selected emails",
+          });
+        }
+      });
+
+      let actionParams = {};
```

Comment:
hmmm you already receive actionParams from the `input`, why do you need to rebuild it? I think it's better if the frontend builds it, and sends it ready on `actionParams` to go into db, but then the validation should also be accurate

what you can do is a type like this:

```
type TriggersForm = {
  action: { action: TriggerAction.SEND_EMAIL, actionParams: { members: string } } | { action: TriggerAction.ADD_TO_DATASET, actionParams: { datasetId: string } }
  ...
}
```

but actually as a zod validator, which you can use to validate the structure both on the react-hook-form on the frontend and here on the endpoint, making it really well tied